packages/riverpod/test/framework/provider_element_test.dart

```diff
@@ -1281,6 +1281,24 @@ void main() {
     });
   });
 
+  group('ref.onReload', () {
+    test('is called on refresh', () {
+      final container = createContainer();
+      final onRefreshListener = OnRefreshListener();
+      when(onRefreshListener()).thenReturn(null);
+      final provider = Provider((ref) {
+        ref.onRefresh(onRefreshListener);
+      });
+      container.read(provider);
+      verifyZeroInteractions(onRefreshListener);
+      container.refresh(provider);
+      verify(onRefreshListener()).called(1);
+      // a second refresh will again call the invalidate listener.
+      container.refresh(provider);
+      verify(onRefreshListener()).called(2);
```

Comment:
That's incorrect. It should be `called(1)`

I believe there's a bug in the current implementation. Listeners are not removed when the provider rebuilds

Reply:
You should clear the list inside _runOnDispose

Reply:
@rrousselGit 
> I believe there's a bug in the current implementation. Listeners are not removed when the provider rebuilds

Could that bug you are taking about be the cause of #1385? Also was there already another issue for this that I didn't find before reporting #1385?

Reply:
oops, you are right, forgot to clear listener list. added reset in `_ronOnDispose` and fixed test.